{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_sequence(sequence_path, to_str=False):\n",
    "    seqs = []\n",
    "    with open(sequence_path, 'r') as f:\n",
    "        for record in SeqIO.parse(f, 'fasta'):\n",
    "            record = SeqRecord(record.seq.upper(), id=record.id, description=record.description)\n",
    "            seqs.append(record)\n",
    "    if to_str:\n",
    "        return [str(seq.seq) for seq in seqs]\n",
    "    else:\n",
    "        return seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "191887\n"
     ]
    }
   ],
   "source": [
    "# 加载cDNA序列\n",
    "cdna_path = './Homo_sapiens.GRCh38.cdna.all.fa'\n",
    "cdna_seq = read_sequence(cdna_path)\n",
    "print(len(cdna_seq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_IDs(line):\n",
    "    gene = None\n",
    "    trans = line.split(\" \")[0]\n",
    "    syml = None\n",
    "    chrom = None\n",
    "    trans_type = None\n",
    "    gene_p = r'gene:(.*?)$'\n",
    "    syml_p = r'gene_symbol:(.*?)$'\n",
    "    chrom_p = r'chromosome:(.*?)$'\n",
    "    trans_p = r'transcript_biotype:(.*?)$'\n",
    "    for elem in line.split(\" \"):\n",
    "        gene_m = re.search(gene_p, elem)\n",
    "        syml_m = re.search(syml_p, elem)\n",
    "        chrom_m = re.search(chrom_p, elem)\n",
    "        trans_m = re.search(trans_p, elem)\n",
    "        if gene_m:\n",
    "            gene = gene_m.group(1)\n",
    "        elif syml_m:\n",
    "            syml = syml_m.group(1)\n",
    "        elif chrom_m:\n",
    "            chrom = chrom_m.group(1)\n",
    "            c = chrom.split(\":\")\n",
    "            chrom = f'{c[1]}:{c[2]}-{c[3]}:{c[4]}'\n",
    "        elif trans_m:\n",
    "            trans_type = trans_m.group(1)\n",
    "    return [trans, gene, syml, chrom, trans_type]\n",
    "\n",
    "def getTransformat(cdna_seq):\n",
    "    anno = []\n",
    "    for record in cdna_seq:\n",
    "        des = record.description\n",
    "        fmt = get_IDs(des)\n",
    "        fmt.append(len(str(record.seq)))\n",
    "        anno.append(fmt)\n",
    "    df = pd.DataFrame(anno, columns=['Transcript_ID', 'Gene_ID', 'Gene_Symbol', 'Gene_Position', 'Transcript_Biotype', 'Transcript_Length'])\n",
    "    df.drop_duplicates(inplace=True)\n",
    "    df.sort_values(by=[\"Gene_Symbol\", \"Transcript_Length\", \"Transcript_ID\", \"Gene_Position\"], ascending=[True, False, True, False], inplace=True)\n",
    "    df.reset_index(drop=True, inplace=True)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Gene_Symbol</th>\n",
       "      <th>Gene_Position</th>\n",
       "      <th>Transcript_Biotype</th>\n",
       "      <th>Transcript_Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000620796.1</td>\n",
       "      <td>ENSG00000275780.1</td>\n",
       "      <td>5S_rRNA</td>\n",
       "      <td>CHR_HSCHR17_2_CTG5:45327366-45327497:-1</td>\n",
       "      <td>rRNA_pseudogene</td>\n",
       "      <td>132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000637405.1</td>\n",
       "      <td>ENSG00000283454.1</td>\n",
       "      <td>5S_rRNA</td>\n",
       "      <td>CHR_HSCHR17_1_CTG5:45327366-45327497:-1</td>\n",
       "      <td>rRNA_pseudogene</td>\n",
       "      <td>132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000648701.1</td>\n",
       "      <td>ENSG00000285609.1</td>\n",
       "      <td>5S_rRNA</td>\n",
       "      <td>1:182944365-182944490:-1</td>\n",
       "      <td>rRNA_pseudogene</td>\n",
       "      <td>126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000648813.1</td>\n",
       "      <td>ENSG00000285674.1</td>\n",
       "      <td>5S_rRNA</td>\n",
       "      <td>19:7886977-7887096:-1</td>\n",
       "      <td>rRNA_pseudogene</td>\n",
       "      <td>120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000647846.1</td>\n",
       "      <td>ENSG00000285776.1</td>\n",
       "      <td>5S_rRNA</td>\n",
       "      <td>X:69672479-69672597:1</td>\n",
       "      <td>rRNA_pseudogene</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191882</th>\n",
       "      <td>ENST00000433749.5</td>\n",
       "      <td>ENSG00000036549.13</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1:77633182-77682675:-1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191883</th>\n",
       "      <td>ENST00000463166.5</td>\n",
       "      <td>ENSG00000036549.13</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1:77633312-77683403:-1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191884</th>\n",
       "      <td>ENST00000474746.1</td>\n",
       "      <td>ENSG00000036549.13</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1:77581818-77585059:-1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191885</th>\n",
       "      <td>ENST00000414381.5</td>\n",
       "      <td>ENSG00000036549.13</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1:77633246-77683374:-1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191886</th>\n",
       "      <td>ENST00000458763.2</td>\n",
       "      <td>ENSG00000036549.13</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1:77676982-77682675:-1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>420</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>191887 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Transcript_ID             Gene_ID Gene_Symbol  \\\n",
       "0       ENST00000620796.1   ENSG00000275780.1     5S_rRNA   \n",
       "1       ENST00000637405.1   ENSG00000283454.1     5S_rRNA   \n",
       "2       ENST00000648701.1   ENSG00000285609.1     5S_rRNA   \n",
       "3       ENST00000648813.1   ENSG00000285674.1     5S_rRNA   \n",
       "4       ENST00000647846.1   ENSG00000285776.1     5S_rRNA   \n",
       "...                   ...                 ...         ...   \n",
       "191882  ENST00000433749.5  ENSG00000036549.13        ZZZ3   \n",
       "191883  ENST00000463166.5  ENSG00000036549.13        ZZZ3   \n",
       "191884  ENST00000474746.1  ENSG00000036549.13        ZZZ3   \n",
       "191885  ENST00000414381.5  ENSG00000036549.13        ZZZ3   \n",
       "191886  ENST00000458763.2  ENSG00000036549.13        ZZZ3   \n",
       "\n",
       "                                  Gene_Position    Transcript_Biotype  \\\n",
       "0       CHR_HSCHR17_2_CTG5:45327366-45327497:-1       rRNA_pseudogene   \n",
       "1       CHR_HSCHR17_1_CTG5:45327366-45327497:-1       rRNA_pseudogene   \n",
       "2                      1:182944365-182944490:-1       rRNA_pseudogene   \n",
       "3                         19:7886977-7887096:-1       rRNA_pseudogene   \n",
       "4                         X:69672479-69672597:1       rRNA_pseudogene   \n",
       "...                                         ...                   ...   \n",
       "191882                   1:77633182-77682675:-1        protein_coding   \n",
       "191883                   1:77633312-77683403:-1  processed_transcript   \n",
       "191884                   1:77581818-77585059:-1  processed_transcript   \n",
       "191885                   1:77633246-77683374:-1        protein_coding   \n",
       "191886                   1:77676982-77682675:-1  processed_transcript   \n",
       "\n",
       "        Transcript_Length  \n",
       "0                     132  \n",
       "1                     132  \n",
       "2                     126  \n",
       "3                     120  \n",
       "4                     119  \n",
       "...                   ...  \n",
       "191882                603  \n",
       "191883                587  \n",
       "191884                572  \n",
       "191885                561  \n",
       "191886                420  \n",
       "\n",
       "[191887 rows x 6 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 将cDNA中的信息提取出来\n",
    "cdna_anno = getTransformat(cdna_seq)\n",
    "cdna_anno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "26\n",
      "         Transcript_ID             Gene_ID Gene_Symbol          Gene_Position  \\\n",
      "0    ENST00000480614.1  ENSG00000112715.24       VEGFA  6:43773172-43786484:1   \n",
      "1    ENST00000372067.7  ENSG00000112715.24       VEGFA  6:43770213-43786487:1   \n",
      "2    ENST00000672860.2  ENSG00000112715.24       VEGFA  6:43770213-43786487:1   \n",
      "3    ENST00000372064.8  ENSG00000112715.24       VEGFA  6:43770707-43786487:1   \n",
      "4    ENST00000372077.8  ENSG00000112715.24       VEGFA  6:43770782-43786487:1   \n",
      "5    ENST00000497139.5  ENSG00000112715.24       VEGFA  6:43777720-43786439:1   \n",
      "6    ENST00000476772.5  ENSG00000112715.24       VEGFA  6:43770184-43778764:1   \n",
      "7    ENST00000425836.6  ENSG00000112715.24       VEGFA  6:43770707-43784949:1   \n",
      "8    ENST00000372055.8  ENSG00000112715.24       VEGFA  6:43770707-43784609:1   \n",
      "9    ENST00000520948.5  ENSG00000112715.24       VEGFA  6:43771083-43784949:1   \n",
      "10   ENST00000413642.7  ENSG00000112715.24       VEGFA  6:43770707-43784589:1   \n",
      "11   ENST00000482630.6  ENSG00000112715.24       VEGFA  6:43770707-43784658:1   \n",
      "12   ENST00000417285.6  ENSG00000112715.24       VEGFA  6:43770707-43784562:1   \n",
      "13  ENST00000230480.10  ENSG00000112715.24       VEGFA  6:43771992-43784935:1   \n",
      "14   ENST00000519767.5  ENSG00000112715.24       VEGFA  6:43770793-43784562:1   \n",
      "15  ENST00000324450.10  ENSG00000112715.24       VEGFA  6:43770707-43784562:1   \n",
      "16   ENST00000523950.5  ENSG00000112715.24       VEGFA  6:43771209-43784902:1   \n",
      "17   ENST00000523873.5  ENSG00000112715.24       VEGFA  6:43771209-43784609:1   \n",
      "18   ENST00000518689.5  ENSG00000112715.24       VEGFA  6:43771247-43784562:1   \n",
      "19   ENST00000518824.5  ENSG00000112715.24       VEGFA  6:43771247-43784658:1   \n",
      "20   ENST00000493786.2  ENSG00000112715.24       VEGFA  6:43778172-43784564:1   \n",
      "21   ENST00000523125.5  ENSG00000112715.24       VEGFA  6:43771247-43784562:1   \n",
      "22   ENST00000518538.5  ENSG00000112715.24       VEGFA  6:43771153-43778937:1   \n",
      "23   ENST00000512683.1  ENSG00000112715.24       VEGFA  6:43771247-43777825:1   \n",
      "24   ENST00000457104.6  ENSG00000112715.24       VEGFA  6:43771247-43784562:1   \n",
      "25   ENST00000520265.1  ENSG00000112715.24       VEGFA  6:43774353-43784562:1   \n",
      "\n",
      "   Transcript_Biotype  Transcript_Length  \n",
      "0     retained_intron              12167  \n",
      "1      protein_coding               3535  \n",
      "2      protein_coding               3535  \n",
      "3      protein_coding               2909  \n",
      "4      protein_coding               2834  \n",
      "5     retained_intron               2746  \n",
      "6     retained_intron               1683  \n",
      "7      protein_coding               1575  \n",
      "8      protein_coding               1286  \n",
      "9      protein_coding               1199  \n",
      "10     protein_coding               1197  \n",
      "11     protein_coding               1146  \n",
      "12     protein_coding               1081  \n",
      "13     protein_coding                993  \n",
      "14     protein_coding                970  \n",
      "15     protein_coding                954  \n",
      "16     protein_coding                954  \n",
      "17     protein_coding                784  \n",
      "18     protein_coding                630  \n",
      "19     protein_coding                606  \n",
      "20    retained_intron                551  \n",
      "21     protein_coding                541  \n",
      "22    retained_intron                535  \n",
      "23    retained_intron                475  \n",
      "24     protein_coding                414  \n",
      "25     protein_coding                336  \n"
     ]
    }
   ],
   "source": [
    "# 将靶标cDNA序列选择出来\n",
    "target_seq = [seq for seq in cdna_seq if 'VEGFA' in seq.description]\n",
    "target_anno = getTransformat(target_seq)\n",
    "print(len(target_seq))\n",
    "print(target_anno)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3\n"
     ]
    }
   ],
   "source": [
    "# 加载sgRNA序列文件\n",
    "sgRNA_path = './sgRNA_sequence.fasta'\n",
    "sgRNA_seq = read_sequence(sgRNA_path)\n",
    "print(len(sgRNA_seq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import pairwise2\n",
    "from Bio.Seq import Seq\n",
    "from tqdm.notebook import tqdm\n",
    "from concurrent.futures import ThreadPoolExecutor\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_cigar_string(alignments, start, end):\n",
    "    cigar = ''\n",
    "    seq1 = alignments[0]\n",
    "    seq2 = alignments[1]\n",
    "    for i in range(start, end):\n",
    "        if seq1[i] == seq2[i]:\n",
    "            cigar += '|'\n",
    "        elif seq1[i] == '-':\n",
    "            cigar += 'I'\n",
    "        elif seq2[i] == '-':\n",
    "            cigar += 'D'\n",
    "        else:\n",
    "            cigar += 'X'\n",
    "    return cigar\n",
    "\n",
    "\n",
    "def pairwise_local_align(seq1, seq2, max_mismatches=None, min_consecutive_matches=None):\n",
    "    try:\n",
    "        alignments = pairwise2.align.localms(seq1.seq, seq2.seq, 2, -1, -8, -0.5, one_alignment_only=True, gap_char='-', force_generic=True)\n",
    "        cigar = get_cigar_string(alignments[0], alignments[0].start, alignments[0].end)\n",
    "        matches = cigar.count('|')\n",
    "        mismatches = cigar.count('X') + cigar.count('I') + cigar.count('D')\n",
    "        start = alignments[0].start\n",
    "        end = alignments[0].end\n",
    "        max_consecutive_matches = max([len(x) for x in re.split('[DXI]', cigar)])\n",
    "        if (max_mismatches and mismatches > max_mismatches) or (min_consecutive_matches and max_consecutive_matches < min_consecutive_matches):\n",
    "            return None\n",
    "        else:\n",
    "            return {'matches': matches, 'mismatches': mismatches, 'start': start, 'end': end,\n",
    "                    'max_consecutive_matches': max_consecutive_matches, \n",
    "                    'seq1':alignments[0][0][start:end], 'seq2':alignments[0][1][start:end], 'cigar':cigar}\n",
    "    except Exception as e:\n",
    "        print(e)\n",
    "        return None\n",
    "    \n",
    "def reverse_complement(seq):\n",
    "    if isinstance(seq, str):\n",
    "        seq = seq.translate(str.maketrans(\"ATCG\", \"TAGC\"))[::-1]\n",
    "    else:\n",
    "        seq.seq = seq.seq.reverse_complement()\n",
    "    return seq\n",
    "    \n",
    "from multiprocessing import Pool\n",
    "import itertools\n",
    "\n",
    "def get_off_targets(cdna_seq, sgRNA_seq, max_mismatches, min_consecutive_matches, num_processes=None,num_threads=None):\n",
    "    off_targets = []\n",
    "    if num_processes:\n",
    "        with Pool(num_processes) as pool:\n",
    "            for sgRNA, cdna in tqdm(itertools.product(sgRNA_seq, cdna_seq), total=len(sgRNA_seq)*len(cdna_seq), desc='sgRNA'):\n",
    "                forward_result = pool.apply_async(pairwise_local_align, args=(cdna, sgRNA, max_mismatches, min_consecutive_matches))\n",
    "                reverse_result = pool.apply_async(pairwise_local_align, args=(cdna, reverse_complement(sgRNA), max_mismatches, min_consecutive_matches))\n",
    "                forward_result = forward_result.get()\n",
    "                reverse_result = reverse_result.get()\n",
    "                if forward_result:\n",
    "                    off_targets.append({'sgRNA': sgRNA.id, 'cdna': cdna.id, 'strand': '+', **forward_result})\n",
    "                if reverse_result:\n",
    "                    off_targets.append({'sgRNA': sgRNA.id, 'cdna': cdna.id, 'strand': '-', **reverse_result})\n",
    "        return pd.DataFrame(off_targets)\n",
    "    elif num_threads:\n",
    "        with ThreadPoolExecutor(max_workers=num_threads) as executor:\n",
    "            for sgRNA, cdna in tqdm(itertools.product(sgRNA_seq, cdna_seq), total=len(sgRNA_seq)*len(cdna_seq), desc='sgRNA'):\n",
    "                forward_result = executor.submit(pairwise_local_align, cdna, sgRNA, max_mismatches, min_consecutive_matches)\n",
    "                reverse_result = executor.submit(pairwise_local_align, cdna, reverse_complement(sgRNA), max_mismatches, min_consecutive_matches)\n",
    "                forward_result = forward_result.result()\n",
    "                reverse_result = reverse_result.result()\n",
    "                if forward_result:\n",
    "                    off_targets.append({'sgRNA': sgRNA.id, 'cdna': cdna.id, 'strand': '+', **forward_result})\n",
    "                if reverse_result:\n",
    "                    off_targets.append({'sgRNA': sgRNA.id, 'cdna': cdna.id, 'strand': '-', **reverse_result})\n",
    "        return pd.DataFrame(off_targets)\n",
    "    else:\n",
    "        for sgRNA, cdna in tqdm(itertools.product(sgRNA_seq, cdna_seq), total=len(sgRNA_seq)*len(cdna_seq), desc='sgRNA'):\n",
    "            forward_result = pairwise_local_align(cdna, sgRNA, max_mismatches, min_consecutive_matches)\n",
    "            reverse_result = pairwise_local_align(cdna, reverse_complement(sgRNA), max_mismatches, min_consecutive_matches)\n",
    "            if forward_result:\n",
    "                off_targets.append({'sgRNA': sgRNA.id, 'cdna': cdna.id, 'strand': '+', **forward_result})\n",
    "            if reverse_result:\n",
    "                off_targets.append({'sgRNA': sgRNA.id, 'cdna': cdna.id, 'strand': '-', **reverse_result})\n",
    "        return pd.DataFrame(off_targets)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(pairwise_local_align(target_seq[17], sgRNA_seq[3]))\n",
    "# for i in pairwise2.align.globalms(\n",
    "#     target_seq[17], sgRNA_seq[3], \n",
    "#     2, -1, -8, -0.5, \n",
    "#     one_alignment_only=True, gap_char='-', force_generic=True\n",
    "#     ):\n",
    "#     print(i)\n",
    "    \n",
    "# for i in pairwise2.align.globalms(\n",
    "#     target_seq[17], reverse_complement(sgRNA_seq[3]), \n",
    "#     2, -1, -8, -0.5, \n",
    "#     one_alignment_only=True, gap_char='-', force_generic=True\n",
    "#     ):\n",
    "#     print(i)\n",
    "    \n",
    "# for i in pairwise2.align.globalxx(\n",
    "#     target_seq[17], sgRNA_seq[3]\n",
    "#     ):\n",
    "#     print(i)\n",
    "#     break\n",
    "    \n",
    "# for i in pairwise2.align.globalxx(\n",
    "#     target_seq[17], reverse_complement(sgRNA_seq[3])\n",
    "#     ):\n",
    "#     print(i)\n",
    "#     break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 使用两两比对获得比对信息\n",
    "# max_mismatches = 6\n",
    "# min_consecutive_matches = 21\n",
    "# threads = 10\n",
    "# off_targets = get_off_targets(\n",
    "#     cdna_seq=cdna_seq, \n",
    "#     sgRNA_seq=sgRNA_seq, \n",
    "#     max_mismatches=max_mismatches, \n",
    "#     min_consecutive_matches=min_consecutive_matches, \n",
    "#     num_processes=threads\n",
    "#     )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def split_sgrna(sgrna_path, out_path, strand=\"both\"):\n",
    "    \"\"\"\n",
    "    将sgRNA的FASTA序列切分为一条序列一个文件，并取反向互补序列\n",
    "    \"\"\"\n",
    "    sglist = []\n",
    "    sgrna_seq = SeqIO.parse(sgrna_path, 'fasta')\n",
    "    for record in sgrna_seq:\n",
    "        record.seq = record.seq.upper()\n",
    "        sgrna_name = record.id.split(\":\")[0]\n",
    "        swater = f'{out_path}_{sgrna_name}.water'\n",
    "\n",
    "        fwd_fa = f'{out_path}_water_sg_fwd_{sgrna_name}.fasta'\n",
    "        fwd_water = f'{out_path}_water_sg_fwd_{sgrna_name}.water'\n",
    "        fmesg = [sgrna_name, record.seq, fwd_fa, fwd_water, swater]\n",
    "\n",
    "        rev_fa = f'{out_path}_water_sg_rev_{sgrna_name}.fasta'\n",
    "        rev_water = f'{out_path}_water_sg_rev_{sgrna_name}.water'\n",
    "        rmesg = [sgrna_name, record.seq.reverse_complement(), rev_fa, rev_water, swater]\n",
    "\n",
    "        time = datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S')\n",
    "        if strand == \"fwd\":\n",
    "            sglist.append(fmesg)\n",
    "            record.seq = record.seq\n",
    "            SeqIO.write(record, fwd_fa, \"fasta\")\n",
    "            print(f'{time} <PARAM>  {fwd_fa} {fmesg[1]}')\n",
    "        elif strand == \"rev\":\n",
    "            sglist.append(rmesg)\n",
    "            record.seq = record.seq.reverse_complement()\n",
    "            SeqIO.write(record, rev_fa, \"fasta\")\n",
    "            print(f'{time} <PARAM>  {rev_fa} {rmesg[2]}')\n",
    "        elif strand == \"both\":\n",
    "            sglist.append(fmesg)\n",
    "            SeqIO.write(record, fwd_fa, \"fasta\")\n",
    "            print(f'{time} <PARAM>  {fwd_fa} {fmesg[1]}')\n",
    "            sglist.append(rmesg)\n",
    "            record.seq = record.seq.reverse_complement()\n",
    "            SeqIO.write(record, rev_fa, \"fasta\")\n",
    "            print(f'{time} <PARAM>  {rev_fa} {rmesg[2]}')\n",
    "    return sglist\n",
    "\n",
    "def get_water_alignment(sgrna_path, cdna_path, out_path, gapopen=10, gapextend=0.5):\n",
    "    \"\"\"\n",
    "    将切分的sgRNA序列与转录本比对\n",
    "    \"\"\"\n",
    "    wcline = WaterCommandline(\n",
    "        gapopen=gapopen,\n",
    "        gapextend=gapextend,\n",
    "        asequence=sgrna_path,\n",
    "        bsequence=cdna_path,\n",
    "        outfile=out_path\n",
    "    )\n",
    "    print(wcline)\n",
    "    stdout, stderr = wcline()\n",
    "    print(stdout + stderr)\n",
    "    return out_path\n",
    "\n",
    "\n",
    "def pool_get_water_alignment(sgrna_list, cdna_path, gapopen=10, gapextend=0.5, threads=2, realign=False):\n",
    "    pool = Pool(processes=threads)\n",
    "    if realign:\n",
    "        params = [(record[2], cdna_path, record[3], gapopen, gapextend) for record in sgrna_list if not os.path.exists(record[3])]\n",
    "    else:\n",
    "        params = [(record[2], cdna_path, record[3], gapopen, gapextend) for record in sgrna_list]\n",
    "    results = []\n",
    "    with tqdm(desc='Aligning', total=len(params)) as pbar:\n",
    "        for param in params:\n",
    "            result = pool.apply_async(func=get_water_alignment, args=param, callback=lambda x: pbar.update())\n",
    "        results.append(result.get())\n",
    "        pool.close()\n",
    "        pool.join()\n",
    "    time = datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S')\n",
    "    print(f'{time} <OUTPUT> {results}')\n",
    "    return results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2023-04-11 17:24:54 <PARAM>  ./Predictions_water_sg_fwd_sgRNA1.fasta GAATGTAGCCATCCATCCTTGTCAA\n",
      "2023-04-11 17:24:54 <PARAM>  ./Predictions_water_sg_rev_sgRNA1.fasta ./Predictions_water_sg_rev_sgRNA1.fasta\n",
      "2023-04-11 17:24:54 <PARAM>  ./Predictions_water_sg_fwd_sgRNA2.fasta ATGTTGGACTCCTCAGTGGGCACACACTCC\n",
      "2023-04-11 17:24:54 <PARAM>  ./Predictions_water_sg_rev_sgRNA2.fasta ./Predictions_water_sg_rev_sgRNA2.fasta\n",
      "2023-04-11 17:24:54 <PARAM>  ./Predictions_water_sg_fwd_sgRNA3.fasta GTGCTGTAGGAAGCTCATCTCTCCTATGTG\n",
      "2023-04-11 17:24:54 <PARAM>  ./Predictions_water_sg_rev_sgRNA3.fasta ./Predictions_water_sg_rev_sgRNA3.fasta\n",
      "[['sgRNA1', Seq('GAATGTAGCCATCCATCCTTGTCAA'), './Predictions_water_sg_fwd_sgRNA1.fasta', './Predictions_water_sg_fwd_sgRNA1.water', './Predictions_sgRNA1.water'], ['sgRNA1', Seq('TTGACAAGGATGGATGGCTACATTC'), './Predictions_water_sg_rev_sgRNA1.fasta', './Predictions_water_sg_rev_sgRNA1.water', './Predictions_sgRNA1.water'], ['sgRNA2', Seq('ATGTTGGACTCCTCAGTGGGCACACACTCC'), './Predictions_water_sg_fwd_sgRNA2.fasta', './Predictions_water_sg_fwd_sgRNA2.water', './Predictions_sgRNA2.water'], ['sgRNA2', Seq('GGAGTGTGTGCCCACTGAGGAGTCCAACAT'), './Predictions_water_sg_rev_sgRNA2.fasta', './Predictions_water_sg_rev_sgRNA2.water', './Predictions_sgRNA2.water'], ['sgRNA3', Seq('GTGCTGTAGGAAGCTCATCTCTCCTATGTG'), './Predictions_water_sg_fwd_sgRNA3.fasta', './Predictions_water_sg_fwd_sgRNA3.water', './Predictions_sgRNA3.water'], ['sgRNA3', Seq('CACATAGGAGAGATGAGCTTCCTACAGCAC'), './Predictions_water_sg_rev_sgRNA3.fasta', './Predictions_water_sg_rev_sgRNA3.water', './Predictions_sgRNA3.water']]\n"
     ]
    }
   ],
   "source": [
    "import datetime\n",
    "# 分割sgRNA到fasta文件\n",
    "out_path = \"./Predictions\"\n",
    "strand = 'both'\n",
    "sgRNA_list = split_sgrna(\n",
    "    sgrna_path=sgRNA_path,\n",
    "    out_path=out_path,\n",
    "    strand=strand\n",
    ")\n",
    "print(sgRNA_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "9c76e6e4e31b4ee996014af06d12483f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Aligning:   0%|          | 0/6 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "water -outfile=./Predictions_water_sg_rev_sgRNA1.water -asequence=./Predictions_water_sg_rev_sgRNA1.fasta -bsequence=./Homo_sapiens.GRCh38.cdna.all.fa -gapopen=10 -gapextend=0.5water -outfile=./Predictions_water_sg_fwd_sgRNA2.water -asequence=./Predictions_water_sg_fwd_sgRNA2.fasta -bsequence=./Homo_sapiens.GRCh38.cdna.all.fa -gapopen=10 -gapextend=0.5water -outfile=./Predictions_water_sg_fwd_sgRNA3.water -asequence=./Predictions_water_sg_fwd_sgRNA3.fasta -bsequence=./Homo_sapiens.GRCh38.cdna.all.fa -gapopen=10 -gapextend=0.5water -outfile=./Predictions_water_sg_rev_sgRNA2.water -asequence=./Predictions_water_sg_rev_sgRNA2.fasta -bsequence=./Homo_sapiens.GRCh38.cdna.all.fa -gapopen=10 -gapextend=0.5water -outfile=./Predictions_water_sg_fwd_sgRNA1.water -asequence=./Predictions_water_sg_fwd_sgRNA1.fasta -bsequence=./Homo_sapiens.GRCh38.cdna.all.fa -gapopen=10 -gapextend=0.5water -outfile=./Predictions_water_sg_rev_sgRNA3.water -asequence=./Predictions_water_sg_rev_sgRNA3.fasta -bsequence=./Homo_sapiens.GRCh38.cdna.all.fa -gapopen=10 -gapextend=0.5\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Smith-Waterman local alignment of sequences\n",
      "\n",
      "Smith-Waterman local alignment of sequences\n",
      "\n",
      "Smith-Waterman local alignment of sequences\n",
      "\n",
      "Smith-Waterman local alignment of sequences\n",
      "\n",
      "Smith-Waterman local alignment of sequences\n",
      "\n",
      "Smith-Waterman local alignment of sequences\n",
      "\n",
      "2023-04-11 17:35:12 <OUTPUT> ['./Predictions_water_sg_rev_sgRNA3.water']\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['./Predictions_water_sg_rev_sgRNA3.water']"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio.Emboss.Applications import WaterCommandline\n",
    "# 批量比对获取比对结果文件列表\n",
    "gapopen = 10\n",
    "gapextend = 0.5\n",
    "threads = 10\n",
    "pool_get_water_alignment(\n",
    "    sgrna_list=sgRNA_list,\n",
    "    cdna_path=cdna_path,\n",
    "    gapopen=gapopen,\n",
    "    gapextend=gapextend,\n",
    "    threads=threads\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import AlignIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cal_mismatch_and_gap(cdna, sgrna, sgrna_len):\n",
    "    cdna_seq = cdna.seq\n",
    "    sgrna_seq = sgrna.seq\n",
    "    unalign = 0\n",
    "    match = 0\n",
    "    max_seed = []\n",
    "    seed = 0\n",
    "    for cdna_base, sgrna_base in zip(cdna_seq, sgrna_seq):\n",
    "        if cdna_base != sgrna_base:\n",
    "            unalign += 1\n",
    "            max_seed.append(seed)\n",
    "            seed = 0\n",
    "        else:\n",
    "            seed += 1\n",
    "            match += 1\n",
    "    max_seed.append(seed)\n",
    "    unmatch = sgrna_len - match\n",
    "    max_seed = max(max_seed)\n",
    "    return match, unmatch, max_seed\n",
    "\n",
    "\n",
    "def filter_water_result(water_out_path, sgrna_len, min_match, max_unalign, sgrna_id):\n",
    "    water = list(AlignIO.parse(water_out_path, 'emboss'))\n",
    "    water_out = []\n",
    "    with tqdm(total=len(water), desc=f'Filtering {sgrna_id}') as pbar:\n",
    "        for record in water:\n",
    "            sgrna, cdna = record\n",
    "            match, unmatch, max_seed_len = cal_mismatch_and_gap(cdna, sgrna, sgrna_len)\n",
    "            align_len = record.get_alignment_length()\n",
    "            if max_seed_len >= 8 and match >= sgrna_len - max_unalign and match >= min_match and sgrna_len + max_unalign >= align_len:\n",
    "                mesg = [sgrna_id, cdna.id, sgrna.seq, cdna.seq, sgrna_len, match, unmatch, max_seed_len, align_len]\n",
    "                water_out.append(mesg)\n",
    "            pbar.update(1)\n",
    "    time = datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S')\n",
    "    print(f'{time} <RESULT> {sgrna_id} got {len(water_out)} results')\n",
    "    return water_out\n",
    "\n",
    "\n",
    "def pool_filter_water_result(sgrna_list, min_match, max_unalign, threads=2):\n",
    "    pool = Pool(processes=threads)\n",
    "    results_list = []\n",
    "    for record in tqdm(sgrna_list, desc='Filtering result'):\n",
    "        water_path = record[3]\n",
    "        sgrna_len = len(record[1])\n",
    "        sgrna_id = record[0]\n",
    "        result = pool.apply_async(filter_water_result, args=(water_path, sgrna_len, min_match, max_unalign, sgrna_id))\n",
    "        results_list.append(result.get())\n",
    "    pool.close()\n",
    "    pool.join()\n",
    "    time = datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S')\n",
    "    print(f'{time} <RESULT> got {len(results_list)} different sgRNA result')\n",
    "    return results_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "699061c9e2794c3bb8aba33547ede346",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Filtering result:   0%|          | 0/6 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2023-04-11 17:37:08 <RESULT> sgRNA1 got 319 results\n",
      "2023-04-11 17:37:18 <RESULT> sgRNA1 got 535 results\n",
      "2023-04-11 17:37:28 <RESULT> sgRNA2 got 385 results\n",
      "2023-04-11 17:37:38 <RESULT> sgRNA2 got 424 results\n",
      "2023-04-11 17:37:49 <RESULT> sgRNA3 got 379 results\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "072c2e477c3648d4b9fbc3d1534d9599",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Filtering sgRNA3:   0%|          | 0/191887 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2023-04-11 17:37:59 <RESULT> sgRNA3 got 610 results\n",
      "2023-04-11 17:37:59 <RESULT> got 6 different sgRNA result\n"
     ]
    }
   ],
   "source": [
    "min_match = 21\n",
    "max_unmatch = 6\n",
    "threads = 5\n",
    "flt_list = pool_filter_water_result(\n",
    "    sgrna_list=sgRNA_list,\n",
    "    min_match=min_match,\n",
    "    max_unalign=max_unmatch,\n",
    "    threads=threads\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge_to_dataframe(multi_list):\n",
    "    alist = []\n",
    "    for res in multi_list:\n",
    "        alist.extend(res)\n",
    "    df = pd.DataFrame(\n",
    "        alist,\n",
    "        columns=['sgRNA_ID', 'Transcript_ID', 'sgRNA_align', \n",
    "                 'Transcript_align', 'sgRNA_len', 'Match', 'Unmatch',\n",
    "                 'Max_seed', 'Alignment_length']\n",
    "    )\n",
    "    df.drop_duplicates()\n",
    "    df.sort_values(\n",
    "        by=[\"sgRNA_ID\", \"Unmatch\", \"Transcript_ID\"], \n",
    "        ascending=[False, True, False], \n",
    "        inplace=True)\n",
    "    df.reset_index(drop=True, inplace=True)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sgRNA_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>sgRNA_align</th>\n",
       "      <th>Transcript_align</th>\n",
       "      <th>sgRNA_len</th>\n",
       "      <th>Match</th>\n",
       "      <th>Unmatch</th>\n",
       "      <th>Max_seed</th>\n",
       "      <th>Alignment_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sgRNA3</td>\n",
       "      <td>ENST00000672860.2</td>\n",
       "      <td>(C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...</td>\n",
       "      <td>(C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sgRNA3</td>\n",
       "      <td>ENST00000523950.5</td>\n",
       "      <td>(C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...</td>\n",
       "      <td>(C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sgRNA3</td>\n",
       "      <td>ENST00000523873.5</td>\n",
       "      <td>(C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...</td>\n",
       "      <td>(C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sgRNA3</td>\n",
       "      <td>ENST00000523125.5</td>\n",
       "      <td>(C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...</td>\n",
       "      <td>(C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sgRNA3</td>\n",
       "      <td>ENST00000520948.5</td>\n",
       "      <td>(C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...</td>\n",
       "      <td>(C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2647</th>\n",
       "      <td>sgRNA1</td>\n",
       "      <td>ENST00000232975.8</td>\n",
       "      <td>(T, T, G, A, C, A, A, G, G, A, T, G, -, -, G, ...</td>\n",
       "      <td>(T, T, G, A, C, A, A, A, A, A, T, G, C, T, G, ...</td>\n",
       "      <td>25</td>\n",
       "      <td>21</td>\n",
       "      <td>4</td>\n",
       "      <td>11</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2648</th>\n",
       "      <td>sgRNA1</td>\n",
       "      <td>ENST00000225742.13</td>\n",
       "      <td>(T, T, G, A, C, A, A, G, G, A, -, -, T, G, G, ...</td>\n",
       "      <td>(T, G, G, A, C, A, A, G, G, A, G, C, T, G, T, ...</td>\n",
       "      <td>25</td>\n",
       "      <td>21</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2649</th>\n",
       "      <td>sgRNA1</td>\n",
       "      <td>ENST00000219789.11</td>\n",
       "      <td>(G, A, A, T, G, T, A, G, C, C, A, T, C, C, A, ...</td>\n",
       "      <td>(G, A, A, -, G, A, A, G, C, C, A, T, C, C, A, ...</td>\n",
       "      <td>25</td>\n",
       "      <td>21</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2650</th>\n",
       "      <td>sgRNA1</td>\n",
       "      <td>ENST00000201963.3</td>\n",
       "      <td>(T, T, G, A, C, A, A, G, G, A, T, G, -, -, -, ...</td>\n",
       "      <td>(T, T, T, A, C, A, T, G, T, A, T, G, A, T, G, ...</td>\n",
       "      <td>25</td>\n",
       "      <td>21</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651</th>\n",
       "      <td>sgRNA1</td>\n",
       "      <td>ENST00000014930.9</td>\n",
       "      <td>(T, G, A, C, A, A, G, G, -, -, A, T, -, -, G, ...</td>\n",
       "      <td>(T, G, A, C, A, A, G, G, G, A, A, T, T, G, G, ...</td>\n",
       "      <td>25</td>\n",
       "      <td>21</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2652 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     sgRNA_ID       Transcript_ID  \\\n",
       "0      sgRNA3   ENST00000672860.2   \n",
       "1      sgRNA3   ENST00000523950.5   \n",
       "2      sgRNA3   ENST00000523873.5   \n",
       "3      sgRNA3   ENST00000523125.5   \n",
       "4      sgRNA3   ENST00000520948.5   \n",
       "...       ...                 ...   \n",
       "2647   sgRNA1   ENST00000232975.8   \n",
       "2648   sgRNA1  ENST00000225742.13   \n",
       "2649   sgRNA1  ENST00000219789.11   \n",
       "2650   sgRNA1   ENST00000201963.3   \n",
       "2651   sgRNA1   ENST00000014930.9   \n",
       "\n",
       "                                            sgRNA_align  \\\n",
       "0     (C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...   \n",
       "1     (C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...   \n",
       "2     (C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...   \n",
       "3     (C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...   \n",
       "4     (C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...   \n",
       "...                                                 ...   \n",
       "2647  (T, T, G, A, C, A, A, G, G, A, T, G, -, -, G, ...   \n",
       "2648  (T, T, G, A, C, A, A, G, G, A, -, -, T, G, G, ...   \n",
       "2649  (G, A, A, T, G, T, A, G, C, C, A, T, C, C, A, ...   \n",
       "2650  (T, T, G, A, C, A, A, G, G, A, T, G, -, -, -, ...   \n",
       "2651  (T, G, A, C, A, A, G, G, -, -, A, T, -, -, G, ...   \n",
       "\n",
       "                                       Transcript_align  sgRNA_len  Match  \\\n",
       "0     (C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...         30     30   \n",
       "1     (C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...         30     30   \n",
       "2     (C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...         30     30   \n",
       "3     (C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...         30     30   \n",
       "4     (C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...         30     30   \n",
       "...                                                 ...        ...    ...   \n",
       "2647  (T, T, G, A, C, A, A, A, A, A, T, G, C, T, G, ...         25     21   \n",
       "2648  (T, G, G, A, C, A, A, G, G, A, G, C, T, G, T, ...         25     21   \n",
       "2649  (G, A, A, -, G, A, A, G, C, C, A, T, C, C, A, ...         25     21   \n",
       "2650  (T, T, T, A, C, A, T, G, T, A, T, G, A, T, G, ...         25     21   \n",
       "2651  (T, G, A, C, A, A, G, G, G, A, A, T, T, G, G, ...         25     21   \n",
       "\n",
       "      Unmatch  Max_seed  Alignment_length  \n",
       "0           0        30                30  \n",
       "1           0        30                30  \n",
       "2           0        30                30  \n",
       "3           0        30                30  \n",
       "4           0        30                30  \n",
       "...       ...       ...               ...  \n",
       "2647        4        11                25  \n",
       "2648        4         8                30  \n",
       "2649        4         9                30  \n",
       "2650        4         8                31  \n",
       "2651        4         8                31  \n",
       "\n",
       "[2652 rows x 9 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "flt_df = merge_to_dataframe(flt_list)\n",
    "flt_file = f'{out_path}_water_fltM{min_match}U{max_unmatch}.result.txt'\n",
    "flt_df.to_csv(flt_file, sep=\"\\t\", index=False, encoding='utf-8')\n",
    "flt_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "def anno_transcript(data=None, data_path=None, trans_anno=None, cdna_path=None):\n",
    "    if data_path and not data:\n",
    "        data =  pd.read_csv(data_path, sep=\"\\t\")\n",
    "    if cdna_path and not trans_anno:\n",
    "        cdna_fa = SeqIO.parse(cdna_path, \"fasta\")\n",
    "        trans_anno = getTransformat(cdna_fa)\n",
    "    data = pd.merge(data, trans_anno, how='inner', on='Transcript_ID')\n",
    "    data.drop_duplicates(inplace=True)\n",
    "    data.sort_values(by=[\"sgRNA_ID\", 'Max_seed',\"Match\"], ascending=[False, False, False], inplace=True)\n",
    "    data.reset_index(drop=True, inplace=True)\n",
    "    return data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sgRNA_ID</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>sgRNA_align</th>\n",
       "      <th>Transcript_align</th>\n",
       "      <th>sgRNA_len</th>\n",
       "      <th>Match</th>\n",
       "      <th>Unmatch</th>\n",
       "      <th>Max_seed</th>\n",
       "      <th>Alignment_length</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Gene_Symbol</th>\n",
       "      <th>Gene_Position</th>\n",
       "      <th>Transcript_Biotype</th>\n",
       "      <th>Transcript_Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sgRNA3</td>\n",
       "      <td>ENST00000672860.2</td>\n",
       "      <td>(C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...</td>\n",
       "      <td>(C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>ENSG00000112715.24</td>\n",
       "      <td>VEGFA</td>\n",
       "      <td>6:43770213-43786487:1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>3535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sgRNA3</td>\n",
       "      <td>ENST00000523950.5</td>\n",
       "      <td>(C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...</td>\n",
       "      <td>(C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>ENSG00000112715.24</td>\n",
       "      <td>VEGFA</td>\n",
       "      <td>6:43771209-43784902:1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sgRNA3</td>\n",
       "      <td>ENST00000523873.5</td>\n",
       "      <td>(C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...</td>\n",
       "      <td>(C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>ENSG00000112715.24</td>\n",
       "      <td>VEGFA</td>\n",
       "      <td>6:43771209-43784609:1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sgRNA3</td>\n",
       "      <td>ENST00000523125.5</td>\n",
       "      <td>(C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...</td>\n",
       "      <td>(C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>ENSG00000112715.24</td>\n",
       "      <td>VEGFA</td>\n",
       "      <td>6:43771247-43784562:1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sgRNA3</td>\n",
       "      <td>ENST00000520948.5</td>\n",
       "      <td>(C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...</td>\n",
       "      <td>(C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "      <td>ENSG00000112715.24</td>\n",
       "      <td>VEGFA</td>\n",
       "      <td>6:43771083-43784949:1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2647</th>\n",
       "      <td>sgRNA1</td>\n",
       "      <td>ENST00000244204.11</td>\n",
       "      <td>(T, T, G, A, C, A, A, G, G, A, T, -, -, -, -, ...</td>\n",
       "      <td>(T, T, G, A, -, A, A, G, G, T, T, A, T, G, A, ...</td>\n",
       "      <td>25</td>\n",
       "      <td>21</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>30</td>\n",
       "      <td>ENSG00000124357.13</td>\n",
       "      <td>NAGK</td>\n",
       "      <td>2:71068648-71079808:1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2648</th>\n",
       "      <td>sgRNA1</td>\n",
       "      <td>ENST00000233055.9</td>\n",
       "      <td>(T, G, A, C, A, A, G, G, A, T, G, -, -, -, -, ...</td>\n",
       "      <td>(T, G, A, C, A, A, G, G, -, T, G, C, A, G, T, ...</td>\n",
       "      <td>25</td>\n",
       "      <td>21</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>30</td>\n",
       "      <td>ENSG00000085449.15</td>\n",
       "      <td>WDFY1</td>\n",
       "      <td>2:223875348-223945335:-1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>4607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2649</th>\n",
       "      <td>sgRNA1</td>\n",
       "      <td>ENST00000225742.13</td>\n",
       "      <td>(T, T, G, A, C, A, A, G, G, A, -, -, T, G, G, ...</td>\n",
       "      <td>(T, G, G, A, C, A, A, G, G, A, G, C, T, G, T, ...</td>\n",
       "      <td>25</td>\n",
       "      <td>21</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>30</td>\n",
       "      <td>ENSG00000108604.16</td>\n",
       "      <td>SMARCD2</td>\n",
       "      <td>17:63832604-63842466:-1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2650</th>\n",
       "      <td>sgRNA1</td>\n",
       "      <td>ENST00000201963.3</td>\n",
       "      <td>(T, T, G, A, C, A, A, G, G, A, T, G, -, -, -, ...</td>\n",
       "      <td>(T, T, T, A, C, A, T, G, T, A, T, G, A, T, G, ...</td>\n",
       "      <td>25</td>\n",
       "      <td>21</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>31</td>\n",
       "      <td>ENSG00000088305.18</td>\n",
       "      <td>DNMT3B</td>\n",
       "      <td>20:32779852-32809356:1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>4255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2651</th>\n",
       "      <td>sgRNA1</td>\n",
       "      <td>ENST00000014930.9</td>\n",
       "      <td>(T, G, A, C, A, A, G, G, -, -, A, T, -, -, G, ...</td>\n",
       "      <td>(T, G, A, C, A, A, G, G, G, A, A, T, T, G, G, ...</td>\n",
       "      <td>25</td>\n",
       "      <td>21</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>31</td>\n",
       "      <td>ENSG00000013583.10</td>\n",
       "      <td>HEBP1</td>\n",
       "      <td>12:12974870-13000265:-1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1159</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2652 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     sgRNA_ID       Transcript_ID  \\\n",
       "0      sgRNA3   ENST00000672860.2   \n",
       "1      sgRNA3   ENST00000523950.5   \n",
       "2      sgRNA3   ENST00000523873.5   \n",
       "3      sgRNA3   ENST00000523125.5   \n",
       "4      sgRNA3   ENST00000520948.5   \n",
       "...       ...                 ...   \n",
       "2647   sgRNA1  ENST00000244204.11   \n",
       "2648   sgRNA1   ENST00000233055.9   \n",
       "2649   sgRNA1  ENST00000225742.13   \n",
       "2650   sgRNA1   ENST00000201963.3   \n",
       "2651   sgRNA1   ENST00000014930.9   \n",
       "\n",
       "                                            sgRNA_align  \\\n",
       "0     (C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...   \n",
       "1     (C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...   \n",
       "2     (C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...   \n",
       "3     (C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...   \n",
       "4     (C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...   \n",
       "...                                                 ...   \n",
       "2647  (T, T, G, A, C, A, A, G, G, A, T, -, -, -, -, ...   \n",
       "2648  (T, G, A, C, A, A, G, G, A, T, G, -, -, -, -, ...   \n",
       "2649  (T, T, G, A, C, A, A, G, G, A, -, -, T, G, G, ...   \n",
       "2650  (T, T, G, A, C, A, A, G, G, A, T, G, -, -, -, ...   \n",
       "2651  (T, G, A, C, A, A, G, G, -, -, A, T, -, -, G, ...   \n",
       "\n",
       "                                       Transcript_align  sgRNA_len  Match  \\\n",
       "0     (C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...         30     30   \n",
       "1     (C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...         30     30   \n",
       "2     (C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...         30     30   \n",
       "3     (C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...         30     30   \n",
       "4     (C, A, C, A, T, A, G, G, A, G, A, G, A, T, G, ...         30     30   \n",
       "...                                                 ...        ...    ...   \n",
       "2647  (T, T, G, A, -, A, A, G, G, T, T, A, T, G, A, ...         25     21   \n",
       "2648  (T, G, A, C, A, A, G, G, -, T, G, C, A, G, T, ...         25     21   \n",
       "2649  (T, G, G, A, C, A, A, G, G, A, G, C, T, G, T, ...         25     21   \n",
       "2650  (T, T, T, A, C, A, T, G, T, A, T, G, A, T, G, ...         25     21   \n",
       "2651  (T, G, A, C, A, A, G, G, G, A, A, T, T, G, G, ...         25     21   \n",
       "\n",
       "      Unmatch  Max_seed  Alignment_length             Gene_ID Gene_Symbol  \\\n",
       "0           0        30                30  ENSG00000112715.24       VEGFA   \n",
       "1           0        30                30  ENSG00000112715.24       VEGFA   \n",
       "2           0        30                30  ENSG00000112715.24       VEGFA   \n",
       "3           0        30                30  ENSG00000112715.24       VEGFA   \n",
       "4           0        30                30  ENSG00000112715.24       VEGFA   \n",
       "...       ...       ...               ...                 ...         ...   \n",
       "2647        4         8                30  ENSG00000124357.13        NAGK   \n",
       "2648        4         8                30  ENSG00000085449.15       WDFY1   \n",
       "2649        4         8                30  ENSG00000108604.16     SMARCD2   \n",
       "2650        4         8                31  ENSG00000088305.18      DNMT3B   \n",
       "2651        4         8                31  ENSG00000013583.10       HEBP1   \n",
       "\n",
       "                 Gene_Position Transcript_Biotype  Transcript_Length  \n",
       "0        6:43770213-43786487:1     protein_coding               3535  \n",
       "1        6:43771209-43784902:1     protein_coding                954  \n",
       "2        6:43771209-43784609:1     protein_coding                784  \n",
       "3        6:43771247-43784562:1     protein_coding                541  \n",
       "4        6:43771083-43784949:1     protein_coding               1199  \n",
       "...                        ...                ...                ...  \n",
       "2647     2:71068648-71079808:1     protein_coding               2371  \n",
       "2648  2:223875348-223945335:-1     protein_coding               4607  \n",
       "2649   17:63832604-63842466:-1     protein_coding               1800  \n",
       "2650    20:32779852-32809356:1     protein_coding               4255  \n",
       "2651   12:12974870-13000265:-1     protein_coding               1159  \n",
       "\n",
       "[2652 rows x 14 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "anno_df = anno_transcript(\n",
    "    data=flt_df, \n",
    "    data_path=None, \n",
    "    trans_anno=cdna_anno, \n",
    "    cdna_path=None)\n",
    "anno_file = f'{out_path}_water_fltM{min_match}U{max_unmatch}.result.annotated.txt'\n",
    "anno_df.to_csv(anno_file, sep=\"\\t\", index=False, encoding='utf-8')\n",
    "anno_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import gzip\n",
    "\n",
    "\n",
    "def remove_tmpfile(sgrna_list):\n",
    "    gzfile = []\n",
    "    time = datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S')\n",
    "    print(f'{time} <RESULT> Creating gzipfile for result and remove tmpfile.')\n",
    "    with tqdm(total=len(sgrna_list), desc='Removing tmpfile') as pbar:\n",
    "        for record in sgrna_list:\n",
    "            sgrna_fa = record[2]\n",
    "            water_out = record[3]\n",
    "            gzipout = f'{water_out}.gz'\n",
    "            if os.path.exists(gzipout):\n",
    "                gzfile.append(gzipout)\n",
    "                pbar.update(1) \n",
    "                continue\n",
    "            with open(water_out, \"rb\") as f_in:\n",
    "                f_value = f_in.read()\n",
    "            f_in.close()\n",
    "\n",
    "            with gzip.open(gzipout, \"wb\", 9) as gzip_out:\n",
    "                gzip_out.write(memoryview(f_value))\n",
    "            gzip_out.close()\n",
    "            try:\n",
    "                os.remove(sgrna_fa)\n",
    "                os.remove(water_out)\n",
    "                gzfile.append(gzipout)\n",
    "            except:\n",
    "                continue\n",
    "            pbar.update(1)        \n",
    "    return gzfile\n",
    "\n",
    "def tar_files(file_list, output_file):\n",
    "    try:\n",
    "        subprocess.run(['tar', '-czvf', output_file] + file_list)\n",
    "        for file in file_list:\n",
    "            os.remove(file)\n",
    "    except:\n",
    "        pass\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2023-04-11 17:37:59 <RESULT> Creating gzipfile for result and remove tmpfile.\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "035fe18004ab437080a5e3d28c8fef12",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Removing tmpfile:   0%|          | 0/6 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['./Predictions_water_sg_fwd_sgRNA1.water.gz', './Predictions_water_sg_rev_sgRNA1.water.gz', './Predictions_water_sg_fwd_sgRNA2.water.gz', './Predictions_water_sg_rev_sgRNA2.water.gz', './Predictions_water_sg_fwd_sgRNA3.water.gz', './Predictions_water_sg_rev_sgRNA3.water.gz']\n",
      "./Predictions_water_sg_fwd_sgRNA1.water.gz\n",
      "./Predictions_water_sg_rev_sgRNA1.water.gz\n",
      "./Predictions_water_sg_fwd_sgRNA2.water.gz\n",
      "./Predictions_water_sg_rev_sgRNA2.water.gz\n",
      "./Predictions_water_sg_fwd_sgRNA3.water.gz\n",
      "./Predictions_water_sg_rev_sgRNA3.water.gz\n",
      "./Predictions_water_fltM21U6.result.tar.gz\n"
     ]
    }
   ],
   "source": [
    "tar_file = f'{out_path}_water_fltM{min_match}U{max_unmatch}.result.tar.gz'\n",
    "gz_files = remove_tmpfile(sgrna_list=sgRNA_list)\n",
    "print(gz_files)\n",
    "tar_files(gz_files, tar_file)\n",
    "print(tar_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Blast.Applications import NcbiblastnCommandline\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio.Seq import Seq\n",
    "from tqdm import tqdm\n",
    "import os\n",
    "\n",
    "def blast_local_align(seq1, seq2, max_mismatches=None, min_consecutive_matches=None):\n",
    "    seq1_file = 'seq1.fasta'\n",
    "    seq2_file = 'seq2.fasta'\n",
    "    with open(seq1_file, 'w') as f:\n",
    "        f.write(f'>{seq1.id}\\n{seq1.seq}\\n')\n",
    "    with open(seq2_file, 'w') as f:\n",
    "        f.write(f'>{seq2.id}\\n{seq2.seq}\\n')\n",
    "    blastn_cline = NcbiblastnCommandline(query=seq1_file, subject=seq2_file, outfmt=5)\n",
    "    stdout, stderr = blastn_cline()\n",
    "    blast_records = NCBIXML.parse(stdout)\n",
    "    for blast_record in blast_records:\n",
    "        for alignment in blast_record.alignments:\n",
    "            for hsp in alignment.hsps:\n",
    "                cigar = ''\n",
    "                seq1 = Seq(hsp.query)\n",
    "                seq2 = Seq(hsp.sbjct)\n",
    "                for i in range(len(seq1)):\n",
    "                    if seq1[i] == seq2[i]:\n",
    "                        cigar += '|'\n",
    "                    elif seq1[i] == '-':\n",
    "                        cigar += 'I'\n",
    "                    elif seq2[i] == '-':\n",
    "                        cigar += 'D'\n",
    "                    else:\n",
    "                        cigar += 'X'\n",
    "                matches = cigar.count('|')\n",
    "                mismatches = cigar.count('X') + cigar.count('I') + cigar.count('D')\n",
    "                max_consecutive_matches = max([len(x) for x in re.split('[DXI]', cigar)])\n",
    "                if (max_mismatches and mismatches > max_mismatches) or (min_consecutive_matches and max_consecutive_matches < min_consecutive_matches):\n",
    "                    return None\n",
    "                else:\n",
    "                    return {'matches': matches, 'mismatches': mismatches, 'start': hsp.query_start, 'end': hsp.query_end,\n",
    "                            'max_consecutive_matches': max_consecutive_matches, \n",
    "                            'seq1':seq1[hsp.query_start-1:hsp.query_end], 'seq2':seq2[hsp.query_start-1:hsp.query_end], 'cigar':cigar}\n",
    "    os.remove(seq1_file)\n",
    "    os.remove(seq2_file)\n",
    "    return None\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import subprocess\n",
    "\n",
    "# # 定义sgRNA序列和全cDNA序列的文件路径\n",
    "# sgRNA_file = \"/mnt/d/download/human_genome/hg38/Homo_sapiens.GRCh38.cdna.all.fa\"\n",
    "# cDNA_file = \"./sgRNA_sequence.fasta\"\n",
    "\n",
    "# # 定义blastn命令行参数\n",
    "# blastn_args = [\"blastn\", \"-query\", sgRNA_file, \"-task\", \"blastn-short\",\n",
    "#                \"-subject\", cDNA_file, \"-outfmt\", \"0\", \"-evalue\", \"0.5\",\n",
    "#                \"-gapopen\", \"7\", \"-gapextend\", \"2\"]\n",
    "\n",
    "# # 运行blastn命令行并将结果保存至文件\n",
    "# with open(\"blast_result.txt\", \"w\") as output_file:\n",
    "#     subprocess.run(blastn_args, stdout=output_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
